### -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- ###
### This file contains the command lines used in R to run the analysed presented in the paper "Sex differentiation and long-distance gene flow in the Antarctic fish Aethotaxis mitopteryx ###
### -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- ### 


		  #################
          ### Filtering ###
          #################

# Set working directory
setwd("path/to/your/working/directory")
# Load required packages
library(ggplot2)


      # -------------------------------------------- #
      # Plots from the first VCFtools filtering step #
      # -------------------------------------------- #

# Plot in R
depth_indv <- read.delim("step01-depth_indv.txt")
ggplot(depth_indv, aes(INDV, MEAN_DEPTH)) + geom_bar(stat="identity", fill = "dodgerblue1", colour = "black", alpha = 0.3) + theme_light() + theme(axis.text.x = element_text(angle=90)) + geom_hline(yintercept=c(5,10), linetype="dashed", color = "red2")

depth_site <- read.delim("step01-depth_site.txt")
summary(depth_site$MEAN_DEPTH)
ggplot(depth_site, aes(MEAN_DEPTH)) + geom_density(fill = "dodgerblue1", colour = "black", alpha = 0.3) + xlim(0,80) + theme_light() + ylab("density")

miss_indv <- read.delim("step01-miss_indv.txt")
ggplot(miss_indv, aes(INDV, F_MISS)) + geom_bar(stat="identity", fill = "dodgerblue1", colour = "black", alpha = 0.3) + theme_light() + theme(axis.text.x = element_text(angle=90)) + scale_y_continuous(breaks=seq(0, 1, 0.1))

miss_site <- read.delim("step01-miss_site.txt")
summary(miss_site$F_MISS)
ggplot(miss_site, aes(F_MISS)) + geom_density(fill = "dodgerblue1", colour = "black", alpha = 0.3) + theme_light() + ylab("density")



      # ----------------------- #
      # Filtering with SNPfiltR #
      # ----------------------- #

library(vcfR)
library(SNPfiltR)

# Load data
# Use data already filtered for GQ, DP, mac 2, and max-missing 0.5
vcfR <- read.vcfR("step02.vcf")
popmap <- read.delim("popmap.txt", header=FALSE)
colnames(popmap) <- c("id", "pop")

### Apply allele balance filter ###
vcfR <- filter_allele_balance(vcfR)


# Visualise missing data per population and sample
missing_by_sample(vcfR=vcfR, popmap=popmap)
missing_by_snp(vcfR)


# Keep SNPs with at least 85% of the data
vcfR <- missing_by_snp(vcfR, cutoff=0.85)
# Check individual situation
missing_by_sample(vcfR=vcfR, popmap=popmap)


  ### Remove invariant sites generated by dropping individuals ###
vcfR <- min_mac(vcfR, min.mac=1)
# Check individual situation
missing_by_sample(vcfR=vcfR, popmap=popmap)


# Save filtered vcf
vcfR::write.vcf(vcfR, "step03.vcf.gz")


  ### Linkage filter to thin SNPs to one per 1000 bp ###
vcfR.thin <- distance_thin(vcfR, min.distance=1000)


# Save filtered vcf
vcfR::write.vcf(vcfR.thin, "step04.vcf.gz")


# -> From here on "step03" and "step04" indicate the name of the filtered SNP dataset used for the different analyses <- #



          ############################
          ### POPULATION STRUCTURE ###
          ############################

      #-----#
      # PCA #
      #-----#

# Set working directory
setwd("path/to/your/working/directory")
# Load required packages
library(vcfR)
library(adegenet)
library(ggplot2)

# Load data
vcf <- read.vcfR("step04.vcf")
# Convert vcfR object to genlight object
x <- vcfR2genlight(vcf)

# Define geographic groups
x$pop <- as.factor(rep(c("AP", "WS"), c(8, 10)))
# Define sex groups
strata(x) <- read.delim("path/to/popmap-sex.txt", header=FALSE)

## Define colours for geographic groups
my.colours <- c("firebrick1", "dodgerblue2")


# Compute PCA
tab.x <- tab(x, freq=TRUE, NA.method="mean")
pca.x <- dudi.pca(tab.x, center=TRUE, scale=FALSE, scannf=FALSE, nf=2)


# Calculate contributions of the displayed principal components to the variance
var.axis.1 <- pca.x$eig[1]/sum(pca.x$eig)*100
var.axis.1
var.axis.2 <- pca.x$eig[2]/sum(pca.x$eig)*100
var.axis.2


# Plot
ggplot(pca.x$li, aes(x=Axis1, y=Axis2)) +
  geom_point(aes(fill=x$pop, shape=x$strata$V2), size=5, alpha=0.8) +
  theme_classic() +
  theme_bw(base_size=18) +
  xlab(paste("PC1 ", "(", round(var.axis.1, 2), " %", ")", sep="")) + 
  ylab(paste("PC2 ", "(", round(var.axis.2, 2), " %", ")", sep="")) +
  scale_fill_manual(name="Populations:",
                    labels=c("East Antarctic Peninsula","East Weddell Sea"),
                    values=my.colours) +
  scale_shape_manual(name="Sex:",
                    labels=c("Female","Male"),
                    values=c(21,22)) +
  theme(legend.position="bottom") +
  guides(fill = guide_legend(override.aes=list(shape=21)))


# Tutorial to controlling legend appearance in ggplot2
# https://aosmith.rbind.io/2020/07/09/ggplot2-override-aes/




    ### Plot scores per SNP to visualise which scaffold weighs more ###

loadings <- cbind(Scf=x$chromosome, Pos=x$position, pca.x$c1)

## Keep only the main scaffolds (the ones corresponding to chromosomes) for plot clarity
# Orginal names
scf_names <- c("HiC_scaffold_1_RagTag","HiC_scaffold_3_RagTag","HiC_scaffold_4_RagTag","HiC_scaffold_5_RagTag",
               "HiC_scaffold_6_RagTag","HiC_scaffold_8_RagTag","HiC_scaffold_10_RagTag","HiC_scaffold_11_RagTag",
               "HiC_scaffold_13_RagTag","HiC_scaffold_21_RagTag","HiC_scaffold_22_RagTag","HiC_scaffold_23_RagTag",
               "HiC_scaffold_24_RagTag","HiC_scaffold_25_RagTag","HiC_scaffold_26_RagTag","HiC_scaffold_28_RagTag",
               "HiC_scaffold_30_RagTag","HiC_scaffold_31_RagTag","HiC_scaffold_32_RagTag","HiC_scaffold_33_RagTag",
               "HiC_scaffold_34_RagTag","HiC_scaffold_35_RagTag","HiC_scaffold_38_RagTag","HiC_scaffold_41_RagTag")
# Subset
loadings <- subset(loadings, loadings$Scf %in% scf_names)


## Replace scaffolds names for plot clarity
new_scf_names <- c("Chr_01","Chr_03","Chr_04","Chr_05","Chr_06",
                   "Chr_08","Chr_10","Chr_11","Chr_13","Chr_21",
                   "Chr_22","Chr_23","Chr_24","Chr_25","Chr_26",
                   "Chr_28","Chr_30","Chr_31","Chr_32","Chr_33",
                   "Chr_34","Chr_35","Chr_38","Chr_41")
# Renaming
loadings$Scf <- new_scf_names[match(loadings$Scf, scf_names)]

# Plot PC1 scores per SNP
ggplot(loadings, aes(x=Pos/1000000)) +
  geom_point(aes(y=CS1, col=Scf)) +
  scale_colour_manual(values=c(rep(c("gray8","gray48"), 12))) +
  theme_classic() +
  theme(legend.position = "none") +
  facet_grid(. ~ Scf, scales="free_x", space="free_x") +
  labs(x ="Position (MB)", y="Scores PC1")

# Plot PC2 scores per SNP
ggplot(loadings, aes(x=Pos/1000000)) +
  geom_point(aes(y=CS2, col=Scf)) +
  scale_colour_manual(values=c(rep(c("gray8","gray48"), 12))) +
  theme_classic() +
  theme(legend.position = "none") +
  facet_grid(. ~ Scf, scales="free_x", space="free_x") +
  labs(x ="Position (MB)", y="Scores PC2")




          ###########
          ### Fst ###
          ###########

# Required package
library(StAMPP)

# Compute Fst with associated p-values
fst.stampp <- stamppFst(x, nboots = 1000, percent = 95, nclusters = 8)

# Look at the results
fst.stampp$Fsts
fst.stampp$Pvalues




          ##############################
          ### Plot Admixture results ###
          ##############################

setwd("path/to/your/working/directory")

# Required packages
library(ggplot2)
library(pophelper)

# Import cross-validation errors
cv <- read.table("step04-cv_errors.txt", quote="\"")
# Plot cross-validation errors
ggplot(data=cv, aes(x=V1, y=V2)) +
  geom_line(linetype = "dashed") +
  geom_point(size=4) +
  theme_bw(base_size=20) +
  theme(panel.grid.minor = element_blank()) +
  scale_x_continuous(name="K", breaks=cv$V1) +
  ylab("Cross-validation error")

# Import Admixture output
slist <- readQ(files=c("path/to/step04.2.Q","path/to/step04_No06.2.Q"))
# Define species groups
cat <- data.frame(Population=rep(c("East Antarctic Peninsula","East Weddell Sea"),
                                 c(8, 10)))
# Plot
plotQ(slist,imgoutput="join", showindlab=F, showlegend=F, showsp=T, showyaxis=T, showticks=T,
      clustercol=c("gray70", "gray15"), barbordercolour="white",
      splab=c("Whole data", "Without scaffold 06"), grplab=cat,
      grplabsize=1.2, indlabsize=3, splabsize=3, panelspacer=0.2,
      exportpath=getwd(), imgtype="pdf")




          #############################
          ### Divergence Statistics ###
          #############################

# Set working directory
setwd("path/to/your/working/directory")
# Load required packages
library(tidyverse)


      #-------------#
      # ΦST and Dxy #
      #-------------#

# Import data
phi.plot <- read.delim("phi-plot.tsv")

## Keep only the main scaffolds (the ones corresponding to chromosomes) for plot clarity
# Original names
scf_names <- c("HiC_scaffold_1_RagTag","HiC_scaffold_3_RagTag","HiC_scaffold_4_RagTag","HiC_scaffold_5_RagTag",
               "HiC_scaffold_6_RagTag","HiC_scaffold_8_RagTag","HiC_scaffold_10_RagTag","HiC_scaffold_11_RagTag",
               "HiC_scaffold_13_RagTag","HiC_scaffold_21_RagTag","HiC_scaffold_22_RagTag","HiC_scaffold_23_RagTag",
               "HiC_scaffold_24_RagTag","HiC_scaffold_25_RagTag","HiC_scaffold_26_RagTag","HiC_scaffold_28_RagTag",
               "HiC_scaffold_30_RagTag","HiC_scaffold_31_RagTag","HiC_scaffold_32_RagTag","HiC_scaffold_33_RagTag",
               "HiC_scaffold_34_RagTag","HiC_scaffold_35_RagTag","HiC_scaffold_38_RagTag","HiC_scaffold_41_RagTag")
# Subset
phi.plot <- subset(phi.plot, phi.plot$Chr %in% scf_names)


## Replace scaffolds names for plot clarity
new_scf_names <- c("Chr_01","Chr_03","Chr_04","Chr_05","Chr_06",
                   "Chr_08","Chr_10","Chr_11","Chr_13","Chr_21",
                   "Chr_22","Chr_23","Chr_24","Chr_25","Chr_26",
                   "Chr_28","Chr_30","Chr_31","Chr_32","Chr_33",
                   "Chr_34","Chr_35","Chr_38","Chr_41")
# Renaming
phi.plot$Chr <- new_scf_names[match(phi.plot$Chr, scf_names)]


# Changing negative values to zero
phi.plot$phi_st <- replace(phi.plot$phi_st, which(phi.plot$phi_st < 0), 0)
phi.plot$Smoothed.Phi_st <- replace(phi.plot$Smoothed.Phi_st, which(phi.plot$Smoothed.Phi_st < 0), 0)
phi.plot$Dxy <- replace(phi.plot$Dxy, which(phi.plot$Dxy < 0), 0)
phi.plot$Smoothed.Dxy <- replace(phi.plot$Smoothed.Dxy, which(phi.plot$Smoothed.Dxy < 0), 0)


# Facet labels
chr_labels <- c("Chr_01"="01","Chr_03"="03","Chr_04"="04","Chr_05"="05","Chr_06"="06",
                "Chr_08"="08","Chr_10"="10","Chr_11"="11","Chr_13"="13","Chr_21"="21",
                "Chr_22"="22","Chr_23"="23","Chr_24"="24","Chr_25"="25","Chr_26"="26",
                "Chr_28"="28","Chr_30"="30","Chr_31"="31","Chr_32"="32","Chr_33"="33",
                "Chr_34"="34","Chr_35"="35","Chr_38"="38","Chr_41"="41")

## ΦST plot ##

ggplot(phi.plot, aes(BP/1000000, phi_st, colour=Chr)) + geom_point(shape=1) +
  scale_colour_manual(values=c(rep(c("gray8","gray48"), 12))) +
  geom_point(aes(BP/1000000, Smoothed.Phi_st), color="springgreen3", shape=16) +
  theme_classic()+
  xlab("Position (MB)") +
  ylab(bquote(Φ[ST])) +
  facet_grid(. ~ Chr, scales="free_x", space="free_x", labeller = as_labeller(chr_labels)) +
  theme(legend.position="none",
        text = element_text(size=20),
        axis.ticks.x = element_blank(),
        axis.text.x = element_blank(),
        axis.title.x = element_blank(),
        axis.title.y = element_text(size=22))


# Only Scaffold 06
phi.plot %>% filter(Chr=="Chr_06") %>%
ggplot(aes(BP/1000000, phi_st)) + geom_point(shape=1) +
  geom_point(aes(BP/1000000, Smoothed.Phi_st), color="springgreen3", shape=19) +
  theme_classic()+
  xlab("Position (MB)") +
  ylab(bquote(Φ[ST])) +
  theme(legend.position = "none")
  

## Dxy plot ##

ggplot(phi.plot, aes(BP/1000000, Dxy, colour=Chr)) + geom_point(shape=1) +
  scale_colour_manual(values=c(rep(c("gray8","gray48"), 12))) +
  geom_point(aes(BP/1000000, Smoothed.Dxy), color="purple", shape=16) +
  theme_classic() +
  xlab("Position (MB)") +
  ylab(expression(italic(D)[XY])) +
  facet_grid(. ~ Chr, scales="free_x", space="free_x", labeller = as_labeller(chr_labels)) +
  theme(legend.position="none",
        text = element_text(size=20),
        axis.ticks.x = element_blank(),
        axis.text.x = element_blank(),
        axis.title.x = element_blank(),
        axis.title.y = element_text(size=22))


# Only Scaffold 06
phi.plot %>% filter(Chr=="Chr_06") %>%
  ggplot(aes(BP/1000000, Dxy)) + geom_point(shape=1) +
  geom_point(aes(BP/1000000, Smoothed.Dxy), color="purple", shape=19) +
  theme_classic()+
  xlab("Position (MB)") +
  ylab(expression(italic(D)[XY])) +
  theme(legend.position = "none")




          #################################
          ### Homozygosity per scaffold ###
          #################################

# Set working directory
setwd("path/to/your/working/directory")

# Import data
hom <- read.csv("step04-homozigosity.txt", sep="")

## Replace scaffolds names for plot clarity
# Original names
scf_names <- c("HiC_scaffold_1_RagTag","HiC_scaffold_3_RagTag","HiC_scaffold_4_RagTag","HiC_scaffold_5_RagTag",
               "HiC_scaffold_6_RagTag","HiC_scaffold_8_RagTag","HiC_scaffold_10_RagTag","HiC_scaffold_11_RagTag",
               "HiC_scaffold_13_RagTag","HiC_scaffold_21_RagTag","HiC_scaffold_22_RagTag","HiC_scaffold_23_RagTag",
               "HiC_scaffold_24_RagTag","HiC_scaffold_25_RagTag","HiC_scaffold_26_RagTag","HiC_scaffold_28_RagTag",
               "HiC_scaffold_30_RagTag","HiC_scaffold_31_RagTag","HiC_scaffold_32_RagTag","HiC_scaffold_33_RagTag",
               "HiC_scaffold_34_RagTag","HiC_scaffold_35_RagTag","HiC_scaffold_38_RagTag","HiC_scaffold_41_RagTag")

# New names
new_scf_names <- c("Chr_01","Chr_03","Chr_04","Chr_05","Chr_06",
                   "Chr_08","Chr_10","Chr_11","Chr_13","Chr_21",
                   "Chr_22","Chr_23","Chr_24","Chr_25","Chr_26",
                   "Chr_28","Chr_30","Chr_31","Chr_32","Chr_33",
                   "Chr_34","Chr_35","Chr_38","Chr_41")
# Renaming
hom$SCF <- new_scf_names[match(hom$SCF, scf_names)]


# Calculate frequency of observed homozygous sites
hom$obs.freq <- hom$O_HOM/hom$N_SITES

# Add sex
females <- read.table("females.txt", quote="\"", comment.char="")
hom$Sex <- ifelse(hom$INDV %in% females$V1, "Female", "Male")

# Facet labels
chr_labels <- c("Chr_01"="01","Chr_03"="03","Chr_04"="04","Chr_05"="05","Chr_06"="06",
                "Chr_08"="08","Chr_10"="10","Chr_11"="11","Chr_13"="13","Chr_21"="21",
                "Chr_22"="22","Chr_23"="23","Chr_24"="24","Chr_25"="25","Chr_26"="26",
                "Chr_28"="28","Chr_30"="30","Chr_31"="31","Chr_32"="32","Chr_33"="33",
                "Chr_34"="34","Chr_35"="35","Chr_38"="38","Chr_41"="41")

# Plot
ggplot(hom, aes(INDV, obs.freq, fill=Sex, shape=Sex)) + geom_point(size=3) +
  scale_fill_manual(values=c(rep(c("#D41159","#1A85FF"),12)), name="Sex:") +
  scale_shape_manual(values=c(21, 22), name="Sex:") +
  theme_classic() +
  ylab("Observed homozygosity") +
  facet_grid(. ~ SCF, scales="free_x", space="free_x", labeller = as_labeller(chr_labels)) +
  theme(legend.position=c(0.5, 0.2),
        legend.background=element_rect(linetype="solid",linewidth=0.5,colour ="black"),
        legend.title = element_text(size=20),
        legend.text = element_text(size=20),
        text = element_text(size=20),
        axis.title.x = element_blank(),
        axis.title.y = element_text(size=22),
        axis.text.x = element_blank(),
        axis.ticks.x = element_blank()) +
		guides(colour = guide_legend(nrow = 1))




          #############
          ### Depth ###
          #############

# Set working directory
setwd("path/to/your/working/directory")
# Load required packages
library(windowscanr)
library(ggplot2)

# Import data
males <- read.csv("males-depth_site.txt", sep="")
females <- read.csv("females-depth_site.txt", sep="")

# Put data together
depth <- males
depth$VAR_DEPTH <- NULL
names(depth)[names(depth) == "MEAN_DEPTH"] <- "Males_depth"
depth$Females_depth <- females$MEAN_DEPTH

## Keep only the main scaffolds (the ones corresponding to chromosomes) for plot clarity
# Original names
scf_names <- c("HiC_scaffold_1_RagTag","HiC_scaffold_3_RagTag","HiC_scaffold_4_RagTag","HiC_scaffold_5_RagTag",
               "HiC_scaffold_6_RagTag","HiC_scaffold_8_RagTag","HiC_scaffold_10_RagTag","HiC_scaffold_11_RagTag",
               "HiC_scaffold_13_RagTag","HiC_scaffold_21_RagTag","HiC_scaffold_22_RagTag","HiC_scaffold_23_RagTag",
               "HiC_scaffold_24_RagTag","HiC_scaffold_25_RagTag","HiC_scaffold_26_RagTag","HiC_scaffold_28_RagTag",
               "HiC_scaffold_30_RagTag","HiC_scaffold_31_RagTag","HiC_scaffold_32_RagTag","HiC_scaffold_33_RagTag",
               "HiC_scaffold_34_RagTag","HiC_scaffold_35_RagTag","HiC_scaffold_38_RagTag","HiC_scaffold_41_RagTag")
# Subsetting
depth <- subset(depth, depth$CHROM %in% scf_names)


## Replace scaffolds names for plot clarity
new_scf_names <- c("Chr_01","Chr_03","Chr_04","Chr_05","Chr_06",
                   "Chr_08","Chr_10","Chr_11","Chr_13","Chr_21",
                   "Chr_22","Chr_23","Chr_24","Chr_25","Chr_26",
                   "Chr_28","Chr_30","Chr_31","Chr_32","Chr_33",
                   "Chr_34","Chr_35","Chr_38","Chr_41")
# Renaming
depth$CHROM <- new_scf_names[match(depth$CHROM, scf_names)]



# Calculate female/male depth ratio
depth$ratio <- depth$Females_depth/depth$Males_depth
mean_ratio <- mean(depth$ratio)
depth$corrected_ratio <- depth$ratio/mean_ratio

quantile(depth$corrected_ratio)
quantile(depth[depth$CHROM == "Chr_06",]$corrected_ratio)


# Calculate average across sliding window
pos_win <- winScan(x=depth, groups="CHROM", position="POS", values="corrected_ratio",
                   win_size=5000000, funs="mean")

quantile(pos_win$corrected_ratio_mean)
quantile(pos_win[pos_win$CHROM == "Chr_06",]$corrected_ratio_mean)


# Facet labels
chr_labels <- c("Chr_01"="01","Chr_03"="03","Chr_04"="04","Chr_05"="05","Chr_06"="06",
                "Chr_08"="08","Chr_10"="10","Chr_11"="11","Chr_13"="13","Chr_21"="21",
                "Chr_22"="22","Chr_23"="23","Chr_24"="24","Chr_25"="25","Chr_26"="26",
                "Chr_28"="28","Chr_30"="30","Chr_31"="31","Chr_32"="32","Chr_33"="33",
                "Chr_34"="34","Chr_35"="35","Chr_38"="38","Chr_41"="41")

# Plot
ggplot(depth, aes(x=POS/1000000, y=corrected_ratio, color=CHROM)) + geom_point() +
  scale_colour_manual(values=c(rep(c("gray8","gray48"), 12))) +
  geom_line(data=pos_win, aes(x=win_mid/1000000, y=corrected_ratio_mean), colour="red", size=2) +
  facet_grid(. ~ CHROM, scales="free_x", space="free_x", labeller = as_labeller(chr_labels)) +
  ylim(0.5, 2.5) +
  labs(x ="Position (Mb)", y="female/male sequencing depth ratio") +
  theme_classic() +
  theme(legend.position="none",
        axis.text.x = element_blank(),
        axis.ticks = element_blank(),
        text = element_text(size=20),
        axis.text.y = element_text(size=20),
        axis.title.x = element_blank(),
        axis.title.y = element_text(size=22),
        panel.border = element_rect(fill=NA))




          ##############################
          ### Linkage disequilibrium ###
          ##############################

# Set working directory
setwd("path/to/your/working/directory")
# Load required packages
library(windowscanr)
library(ggplot2)

# Import data
ld <- read.csv("step03.ld", sep="")


## Keep only the main scaffolds (the ones corresponding to chromosomes) for plot clarity
# Original names
scf_names <- c("HiC_scaffold_1_RagTag","HiC_scaffold_3_RagTag","HiC_scaffold_4_RagTag","HiC_scaffold_5_RagTag",
               "HiC_scaffold_6_RagTag","HiC_scaffold_8_RagTag","HiC_scaffold_10_RagTag","HiC_scaffold_11_RagTag",
               "HiC_scaffold_13_RagTag","HiC_scaffold_21_RagTag","HiC_scaffold_22_RagTag","HiC_scaffold_23_RagTag",
               "HiC_scaffold_24_RagTag","HiC_scaffold_25_RagTag","HiC_scaffold_26_RagTag","HiC_scaffold_28_RagTag",
               "HiC_scaffold_30_RagTag","HiC_scaffold_31_RagTag","HiC_scaffold_32_RagTag","HiC_scaffold_33_RagTag",
               "HiC_scaffold_34_RagTag","HiC_scaffold_35_RagTag","HiC_scaffold_38_RagTag","HiC_scaffold_41_RagTag")
# Excluding
ld <- subset(ld, ld$CHR_A %in% scf_names)


## Replace scaffolds names for plot clarity
new_scf_names <- c("Chr_01","Chr_03","Chr_04","Chr_05","Chr_06",
                   "Chr_08","Chr_10","Chr_11","Chr_13","Chr_21",
                   "Chr_22","Chr_23","Chr_24","Chr_25","Chr_26",
                   "Chr_28","Chr_30","Chr_31","Chr_32","Chr_33",
                   "Chr_34","Chr_35","Chr_38","Chr_41")
# Renaming
ld$CHR_A <- new_scf_names[match(ld$CHR_A, scf_names)]


  # Sliding window #

# Calculate distance
ld$dist <- abs(ld$BP_A - ld$BP_B)
# Sort by distance per chromosome
# with base R
ld <- ld[order(ld$CHR_A, ld$dist),]
# with dplyr
#library(dplyr)
#ld <- group_by(ld, CHR_A) %>% arrange(dist, .by_group=T)

# Calculate average across sliding window
pos_win <- winScan(x=ld, groups="CHR_A", position="dist", values="R2",
                   win_size=10000, win_step=10000, funs="mean")


# Facet labels
chr_labels <- c("Chr_01"="01","Chr_03"="03","Chr_04"="04","Chr_05"="05","Chr_06"="06",
                "Chr_08"="08","Chr_10"="10","Chr_11"="11","Chr_13"="13","Chr_21"="21",
                "Chr_22"="22","Chr_23"="23","Chr_24"="24","Chr_25"="25","Chr_26"="26",
                "Chr_28"="28","Chr_30"="30","Chr_31"="31","Chr_32"="32","Chr_33"="33",
                "Chr_34"="34","Chr_35"="35","Chr_38"="38","Chr_41"="41")

# Plot
ggplot(pos_win, aes(x=win_mid/1000, y=R2_mean)) + geom_point() +
  facet_wrap(~ CHR_A, ncol=4, labeller = as_labeller(chr_labels)) +
  labs(x ="Distance (Kb)", y=expression(italic(r)^2)) +
  theme(text = element_text(size=25),
        axis.title.x = element_text(size=25),
        axis.title.y = element_text(size=25),
        plot.margin=margin(r=18))


  ## Linkage analysis divided by sex

# Repeat all the previous steps (except plotting) but using as input file the sex-specific files
#ld <- read.csv("step03-males.ld", sep="")
#ld <- read.csv("step03-females.ld", sep="")

# When calculating the average across sliding window, remember to give different names to male/female pos_win resulting object
# male <- pos_win
# female <- pos_win

# Add a column with sex for all the rows
male$sex <- rep("male", dim(male)[1])
female$sex <- rep("female", dim(female)[1])

# Merge the two tables
pos_win <- rbind(male, female)

# Plot
ggplot(pos_win, aes(x=win_mid/1000, y=R2_mean, col=sex)) + geom_point(size=3) +
  scale_color_manual(values=c("#D41159","#1A85FF"), name="Sex:") +
  facet_wrap(~ CHR_A, ncol=4, labeller = as_labeller(chr_labels)) +
  labs(x ="Distance (Kb)", y=expression(italic(r)^2)) +
  theme(text = element_text(size=25),
        axis.title.x = element_text(size=25),
        axis.title.y = element_text(size=25),
        plot.margin=margin(r=18),
        legend.position="bottom",
        legend.title = element_text(size=25),
        legend.text = element_text(size=25))




    ##############################
    # PCA without HiC_scaffold_6 #
    ##############################

setwd("path/to/your/working/directory")
# Load required packages
library(vcfR)
library(adegenet)
library(ggplot2)

# Load data
vcf <- read.vcfR("step04.vcf")
# Convert vcfR object to genlight object
x <- vcfR2genlight(vcf)

# Define species groups
x$pop <- as.factor(rep(c("AP", "WS"), c(8, 10)))
# Define sex groups
strata(x) <- read.delim("path/to/popmap-sex.txt", header=FALSE)

## Define colours for geographic groups
my.colours <- c("firebrick1", "dodgerblue2")

# Exclude HiC_scaffold_6
x <- x[1:18, x$chromosome != "HiC_scaffold_6_RagTag"]


# Compute PCA
tab.x <- tab(x, freq=TRUE, NA.method="mean")
pca.x <- dudi.pca(tab.x, center=TRUE, scale=FALSE, scannf=FALSE, nf=2)

# Calculate contributions of the displayed principal components to the variance
var.axis.1 <- pca.x$eig[1]/sum(pca.x$eig)*100
var.axis.1
var.axis.2 <- pca.x$eig[2]/sum(pca.x$eig)*100
var.axis.2


# Plot populations
ggplot(pca.x$li, aes(x=Axis1, y=Axis2)) +
  geom_point(aes(fill=x$pop, shape=x$strata$V2), size=5, alpha=0.8) +
  theme_classic() +
  theme_bw(base_size=18) +
  xlab(paste("PC1 ", "(", round(var.axis.1, 2), " %", ")", sep="")) + 
  ylab(paste("PC2 ", "(", round(var.axis.2, 2), " %", ")", sep="")) +
  scale_fill_manual(name="Populations:",
                    labels=c("East Antarctic Peninsula","East Weddell Sea"),
                    values=my.colours) +
  scale_shape_manual(name="Sex:",
                     labels=c("Female","Male"),
                     values=c(21,22)) +
  theme(legend.position="bottom") +
  guides(fill = guide_legend(override.aes=list(shape=21))) +
  theme(legend.position="bottom")




          ##############
          ### RADsex ###
          ##############

setwd("path/to/your/working/directory")

# Plot distribution
pdf("distrib_1.pdf", width=4.3, height=5.3)
sgtr::radsex_distrib("distrib_1.tsv", title="Minimum depth threshold: 1", group_labels=c("males", "females"))
dev.off()

pdf("distrib_2.pdf", width=4.3, height=5.3)
sgtr::radsex_distrib("distrib_2.tsv", title = "Minimum depth threshold: 2", group_labels = c("males", "females"))
dev.off()

pdf("distrib_5.pdf", width=4.3, height=5.3)
sgtr::radsex_distrib("distrib_5.tsv", title = "Minimum depth threshold: 5", group_labels = c("males", "females"))
dev.off()

pdf("distrib_10.pdf", width=4.3, height=5.3)
sgtr::radsex_distrib("distrib_10.tsv", title = "Minimum depth threshold: 10", group_labels = c("males", "females"))
dev.off()



# Make circos plot
# Circos plot are plotted with the function sgtr::radsex_map_circos.
# This function requires that all chromosomes present in the 'map_results.tsv' produced by 'radsex map' are also present in the file listing the chromosome names ('chromosomes_names.tsv' in the RADsex tutorial)

# Import mapped markers produced by radsex map
map <- read.delim("D:/aetho/radsex_results/depth_10/map_10.tsv", comment.char="#")

## Keep only the main scaffolds (the ones corresponding to chromosomes) for plot clarity
# Original names
scf_names <- c("HiC_scaffold_1_RagTag", "HiC_scaffold_3_RagTag", "HiC_scaffold_4_RagTag", "HiC_scaffold_5_RagTag",
               "HiC_scaffold_6_RagTag", "HiC_scaffold_8_RagTag", "HiC_scaffold_10_RagTag", "HiC_scaffold_11_RagTag",
			   "HiC_scaffold_13_RagTag", "HiC_scaffold_21_RagTag", "HiC_scaffold_22_RagTag", "HiC_scaffold_23_RagTag",
			   "HiC_scaffold_24_RagTag", "HiC_scaffold_25_RagTag", "HiC_scaffold_26_RagTag", "HiC_scaffold_28_RagTag",
			   "HiC_scaffold_30_RagTag", "HiC_scaffold_31_RagTag", "HiC_scaffold_32_RagTag", "HiC_scaffold_33_RagTag",
			   "HiC_scaffold_34_RagTag", "HiC_scaffold_35_RagTag", "HiC_scaffold_38_RagTag", "HiC_scaffold_41_RagTag")

# Subset
map <- subset(map, map$Contig %in% scf_names)

# Save the file (sgtr::radsex_map_circos works by reading this file so we need to write it)
write.table(map, file="modified_map_for_plotting_10.txt", quote=F, sep="\t", dec=".", na="NA", row.names=F, col.names=T)
# Copy the first line starting with the "#" in the orginal output from  'radsex map' and paste it as first line in the filtered file.

# Make circos plot
sgtr::radsex_map_circos("modified_map_for_plotting_10.txt", chromosomes_file = "chr_ragtag_radsex.txt", tracks=c("bias","p"), output_file = "map_10.png")




          ###########
          ### Map ###
          ###########

library(ggOceanMaps)
library(ggspatial)
library(ggrepel)

Area <- c("East Weddell Sea", "East Antarctic Peninsula")
Latitude <- c(-72.2965, -63.70074)
Longitude <- c(-16.84717, -57.03362)
text_lat <- c(-72, -64.7)
text_lon <- c(-19, 50)

coord <- data.frame(Area, Latitude, Longitude, text_lat, text_lon)
my.colours <- c("firebrick1", "dodgerblue2")

basemap(limits = c(-63, -17, -60, -77), rotate = TRUE, base_size=15) +
  geom_spatial_point(data = coord, aes(x=Longitude, y=Latitude, fill=Area), shape=21, size=5) +
  scale_fill_manual(values=my.colours) +
  theme(legend.position="none") +
  geom_spatial_label(data=coord, aes(x=text_lon, y=text_lat, label=Area), size=5)
